### ISSUE TITLE
Ask Some Questions

### ISSUE THREAD

**Issue by N-Kingsley** [ratschlab/SOM-VAE#1]
Hi,

I have same question for your code. When I run the code two epoches, it works and just can't reconstruct clearly. But when I run five epoches, reconstruction has been distorted. Shouldn't it be more epoches, more clearly?

**Comment by fortuin** [comment]
Thanks for your question! Could you tell us which hyperparameters you are using? The default ones? And are you training on MNIST or a different data set?

**Comment by N-Kingsley** [comment]
The default hyperparameters and MNIST dataset.

**Comment by fortuin** [comment]
I would guess that during the first epochs, the optimization might still oscillate a bit between the different parts of the objective function, since the learning rate is still pretty high. In our experiments, we usually trained for at least 20 epochs (which is also the default value in our code). With our learning rate decay, the optimization should become more balanced later on. Could you maybe try training for 20 epochs or more and report your observations?

**Comment by N-Kingsley** [comment]
I trained for 20 epoched.But still can't reproduce clearly.T
The performance is better when I trained for only 2 epoches. Why?

**Comment by N-Kingsley** [comment]
How about your results?

**Comment by N-Kingsley** [comment]
This time, model trained for 2 epoches doesn't perform well. It may take a lot of training to get good performance.

**Comment by fortuin** [comment]
Hmm, I see... Since it is SGD with some random initialization, the results will surely be subject to some amount of randomness. But I think that it should be possible to reproduce the results in principle. A few things to try to get closer to our MNIST experiment in the paper would be to

- Change the architecture from SOM-VAE-prob to just the SOM-VAE by setting `gamma` and `tau` to 0. We don't expect consecutive MNIST digits to have any time series nature, so we don't need the two time-series-specific losses.
- Make the learning rate decay a bit less quickly, by setting the decay rate to something like 0.95.
- Use a SOM-size of 16 embeddings, that is, 4 by 4. That is the size we used in our clustering experiments.

I hope this helps!

**Comment by N-Kingsley** [comment]
In original paper, the embedding was set to 512. And I want to ask how to define neighbors in embedding.
512 has no integer square root.

**Comment by fortuin** [comment]
Sorry for the confusion. As can be seen in the supplementary figures of our paper, the SOM had a size of 32 by 16 (32*16=512) in that case.

**Comment by N-Kingsley** [comment]
Thanks.  Do you mean this is a matirx of 32 row and 16 column?
And in 'somvae_model.py',  loss = (self.loss_reconstruction + self.alpha*self.loss_commit + self.beta*self.loss_som + self.tau*self.loss_z_prob). Why not add 'self.gamma*loss_probabilities' mentioned in paper?

**Comment by fortuin** [comment]
Yes, 32 rows and 16 columns.
Thanks for the observation with the loss, that was a good catch! I added the probability loss to the sum now.

**Comment by N-Kingsley** [comment]
Thanks. 
I have another question that whether this model can be used to classify different frames in atari games, such as 'Pong-v0' or 'mspacman-v0'. In this case, should I remove the pooling layer?

**Comment by fortuin** [comment]
We have never tried anything like that, but would be really interested to hear about your results! I think I would keep the pooling layer, but you can try different architectures and see which one works best.
